### ISSUE TITLE
Need a DOI before the release

### ISSUE THREAD

**Issue by richardreeve** [BES2016Workshop/CodePublication-and-citation#6]
There's a logical failure in making a DOI using Zenodo - I have to make the release before Zenodo will allocate a DOI. This means that the DOI cannot, by definition, be part of the release. So if someone downloads the release from github (not via Zenodo, natch), then they won't get the DOI citation information (or even worse they'll get out of date citation information!) with it. This seems to significantly reduce the value of having a DOI for the release.

A sensible workaround would be to *preallocate the DOI for the next release* for every enabled repository, then it could be included with the release along with the badge (which would logically be a differently coloured or annotated DOI graphic - or even just a broken link - until the release happened).

**Comment by AugustT** [comment]
Just add the DOI information to the release notes for your release on Github? Seems like an easy fix.

If you really need the DOI before release (I don't think you do), use a different fork on github to mint the DOI

**Comment by richardreeve** [comment]
Thanks - there are lots of partial workarounds, but unless there's an easy way of writing an R package so that citation("pkg") gives the DOI of the package along with the other information, then it's not really solving my problem. I think preallocating DOIs is cheap, easy and would provide useful additional functionality to Zenodo.

I don't think either of your solutions work in that respect (and if I understand your second solution correctly - and Zenodo! - I don't think it works at all as it'll give a DOI for the fork, not for the core repository).

**Comment by AugustT** [comment]
I'm not sure I follow the issue. If you get a doi for your package you can add that to your citation for your package. I think the issue you are raising is that in a strict sense by changing the citation in the package it no longer matches the doi version? Is that right?

**Comment by richardreeve** [comment]
Yes, exactly. The whole point of Zenodo is that it gives a DOI to a specific tag associated with a specific release of the package. The tagged version of *myPackage* - say v1.2.1 - that is assigned a specific DOI - say 12312.12/123424534 - can therefore by definition never know or refer to its DOI as the DOI won't have been set at tag time. Therefore when someone installs the package, and types:

```{R}
> citation("myPackage")
```

the package can never respond with its DOI... unless the DOI is pre-allocated. It will have to choose between referring to an out of date DOI for the previous release or not providing a DOI at all, which will significantly the value of either Zenodo or R's `citation()` command or both....

**Comment by AugustT** [comment]
I think I would just add the DOI to the citation in the package and not worry that that means the package and DOI version are not 100% identical. I think this is an ideological issue rather than one with any practical ramifications.

**Comment by richardreeve** [comment]
I don't think calling it ideological is remotely helpful or fair. The DOI refers to a verifiable release and is only created after it, therefore by definition the package can't have the DOI in it because it was created at release time, not at the later DOI creation time.

It's conceivable you believe we should start tampering with the git repository (which is easy enough) in order to create a second release with the same tag in order to retrospectively add in the DOI, but then, quite apart from it breaking the whole point of Zenodo (of giving a DOI to a specific verifiable release), if Zenodo is remotely reliable it should identify a specific git commit (or some other fingerprint of the tagged code), not to a tag itself, so will now refer to a non-existent git commit, which would provide evidence of our fraud.

I think we stand on different sides of the verifiable releases argument, which I believe I'm trying to improve, and I don't think we're going to get any closer by arguing about it on github.

Cheers,

Richard.

**Comment by AugustT** [comment]
Thanks for a bit of healthy debate! Tom

**Comment by richardreeve** [comment]
Likewise!

**Comment by AugustT** [comment]
Just found you can in fact reserve a DOI on Zenodo! Promblem solved I think 

https://zenodo.org/deposit/new

In this uploads page there is a button 'Pre-reserve DOI'

Happy days

**Comment by richardreeve** [comment]
Yes, that's great. Thanks for spotting it. It does look like you can reserve a DOI for an upload, but it's not clear that you can tie that DOI to a github release - that still seems to generate a new DOI... still, good news that they do it - now we just need it all tied together.
